core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Example.scala

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.concurrent.Future
+
+import akka.Done
+import com.softwaremill.react.kafka_api_producer.Producer
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+
+trait Example {
+  implicit val ec = scala.concurrent.ExecutionContext.Implicits.global
+
+  def consumerProvider: () => KafkaConsumer[Array[Byte], String] = ???
+
+  def producerProvider: () => KafkaProducer[Array[Byte], String] = ???
+
+  class DB {
+    def save(offset: Long, data: String): Future[Done] = ???
+  }
+}
+
+// Consume messages and store a representation, including offset, in DB
+object Example3 extends Example {
+  val db = new DB
+
+  Consumer.source(consumerProvider)
```

Comment:
> stage should clean up resources on completion

Yes, on *stage* completion, but that does not mean that the stage cannot live after all of its ports has been closed!. In fact, how would a Sink, wrapping an API with an async `flush()` method work after its input has been closed (since that is the point where it must flush)? In fact there is a method called `setKeepGoing` which allows the stage to stay alive even after all ports have been closed.

> indication of completion is onComplete or onFailure on all outputs

... and `cancel` for inputs. These signal that no more elements (request) will flow, but has nothing to do with whether the stage has finished or not. In fact, we have always warned people that just because a Sink.file received an onComplete is in no way indicator that the file has been closed, this is why these Sinks materialize to a Future which signals actual completion.

> This allows for example easily have endless Source[Source[Source[T]]] and be sure about resources usage, because all I have to do is to control number of simultaneous working instances of Source

This is already not true even if we don't consider stages. Imagine the following setup (pseudocode):

```scala
val source Graph.create {
   infiniteSource ~> bcast ~> someSink
   bcast ~> take(10) ~> out
   SourceShape(out)
}
```

The above Source will finish its output much earlier than the composite Source actually finishes (which will only happen if `someSink` cancels).

> You may end up with a message that have no chance to be completed and you have no way to be notified that your client is closed.

I don't understand this part. How is this different from the immediate onComplete? The messages which are still flowing downstream while the onComplete starts to propagate further upstream are already non-committable as the Source is no longer alive in your model. This is exactly what is mitigated (up to a timeout) by my proposal.